### DISCUSSION TITLE
How to make `ndarray` default backend container instead of array in rhai?

### DISCUSSION THREAD

**Discussion by rawhuul** [rhaiscript/rhai#927] [Q&A] [answered]
Hey, how to make all array backed by `ndarray` data structure. I've looked into the book and tried to achieve the same using custom syntax, which is as follows:

```rust
engine.register_custom_syntax(
        ["var", "$ident$", "=", "$expr$"],
        true,
        |context, inputs| {
            let var_name = inputs[0].get_string_value().unwrap();
            let expr = &inputs[1];

            // Evaluate the expression
            let value = context.eval_expression_tree(expr)?;

            // Push a new variable into the scope if it doesn't already exist.
            // Otherwise just set its value.
            if !context.scope().is_constant(var_name).unwrap_or(false) {
                if value.is_array() {
                    let series = NdArrayWrapper::from(value);

                    context.scope_mut().set_value(var_name, series);
                } else {
                    context.scope_mut().set_value(var_name, value);
                }

                Ok(Dynamic::UNIT)
            } else {
                Err(format!("Variable `{var_name}` is declared as constant",)
                    .into())
            }
        },
    )?;
```

It works great but only when used as 
```js
var a = [1, 2, 3]
```

not as standalone array
```js
[1, 2, 3]
```

How can I achieve this?

**Comment by schungx** [comment, marked as answer]
It would be quite difficult to override the standard arrays in Rhai, as arrays have lots of built-in handling.

You can consider using a prefix, such as:
```rust
let x = ND#[1, 2, 3];
```
In that case, the custom syntax starts with `ND#` and the value returned is an `NdArray`.

**Reply by rawhuul** [reply]
So, the only workaround involves using custom syntax. Btw, if there's a difference between rhai's `Array` and `Vec<Dynamic>` types.

**Reply by schungx** [reply]
`Array` is defined as `Vec<Dynamic>`...  so there is no difference in the types.

`ndarray` can only hold numbers, so it is more restrictive than Rhai's array.

**Reply by rawhuul** [reply]
Yah, got it

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.
